src/c_string.rs

```diff
@@ -0,0 +1,594 @@
+    /// if the bytes contain an interior nul byte.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use heapless::CString;
+    ///
+    /// let mut c_string = CString::<10>::new();
+    ///
+    /// c_string.extend_from_bytes(b"hey").unwrap();
+    /// c_string.extend_from_bytes(b" there\0").unwrap();
+    ///
+    /// assert_eq!(c_string.to_str(), Ok("hey there"));
+    /// ```
+    pub fn extend_from_bytes(&mut self, bytes: &[u8]) -> Result<(), ExtendError> {
+        let Some(capacity) = self.capacity_with_bytes(bytes) else {
+            return Ok(());
+        };
+
+        if capacity > N {
+            // Cannot store these bytes due to an insufficient capacity.
+            return Err(CapacityError.into());
+        }
+
+        match CStr::from_bytes_with_nul(bytes) {
+            Ok(_) => {
+                // SAFETY: A string is left in a valid state because appended bytes are nul-terminated.
+                unsafe { self.extend_from_bytes_unchecked(bytes) }?;
+
+                Ok(())
+            }
+            Err(FromBytesWithNulError::InteriorNul { position }) => {
+                Err(ExtendError::InteriorNul { position })
+            }
+            Err(FromBytesWithNulError::NotNulTerminated) => {
+                // Because given bytes has no nul byte anywhere, we insert the bytes and
+                // then add the nul byte terminator.
+                //
+                // We've ensured above that we have enough space left to insert these bytes,
+                // so the operations below must succeed.
+                //
+                // SAFETY: We append a missing nul terminator right below.
+                unsafe {
+                    self.extend_from_bytes_unchecked(bytes).unwrap();
+                    self.inner.push_unchecked(0);
+                };
+
+                Ok(())
+            }
+        }
+    }
+
+    /// Removes the nul byte terminator from the inner buffer.
+    ///
+    /// # Safety
+    ///
+    /// Callers must ensure to add the nul terminator back after this function is called.
+    #[inline]
+    unsafe fn pop_terminator(&mut self) {
+        debug_assert_eq!(self.inner.last(), Some(&0));
+
+        // SAFETY: We always have the nul terminator at the end.
+        unsafe { self.inner.pop_unchecked() };
+    }
+
+    /// Removes the existing nul terminator and then extends `self` with the given bytes.
+    ///
+    /// # Safety
+    ///
+    /// If `additional` is not nul-terminated, the [`CString`] is left non nul-terminated, which is
+    /// an invalid state. Caller must ensure that either `additional` has a terminating nul byte
+    /// or ensure to append a trailing nul terminator.
+    unsafe fn extend_from_bytes_unchecked(
+        &mut self,
+        additional: &[u8],
+    ) -> Result<(), CapacityError> {
+        self.pop_terminator();
+
+        self.inner.extend_from_slice(additional)
```

Comment:
Add a nested `unsafe` and document why it is safe.

Reply:
This is a safe method, I think...

Reply:
Right, `inner` is a `Vec`.